### PR TITLE
Provide support for selective role sharing with application sharing

### PR DESCRIPTION
**Pull request by sahandilshan**
## Purpose
$subject

## TODOs:
- Implementation for Get shared application with hierarchical order
- Need to test out edge cases
- Need to add unit tests

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 1]
```diff
@@ -45,6 +50,50 @@ public interface OrgApplicationManager {
     void shareOrganizationApplication(String ownerOrgId, String mainAppId, boolean shareWithAllChildren,
                                       List<String> sharedOrgs) throws OrganizationManagementException;
 
+    default void selectiveShareOrganizationApplication(String applicationId,
+                                                       List<SelectiveApplicationShare> selectiveApplicationShareList)
+            throws OrganizationManagementException {
+        throw new NotImplementedException(
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:56]:
```suggestion
            throws OrganizationManagementException {
            
        throw new NotImplementedException(
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrganizationShareProcessor.java [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Utility class to process organization shares ordered by hierarchy.
+ */
+public class OrganizationShareProcessor {
+
+    // Helper structure to hold results from BFS traversal
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrganizationShareProcessor.java:40]:
```suggestion
    // Helper structure to hold results from BFS traversal.
```

File: components/org.wso2.carbon.identity.organization.management.application/pom.xml [hunk 3]
```diff
@@ -207,13 +211,19 @@
...
+                            org.wso2.carbon.identity.organization.resource.sharing.policy.management.model;version="${org.wso2.identity.organization.mgt.imp.pkg.version.range}",
+                            org.wso2.carbon.identity.organization.resource.sharing.policy.management.util;version="${org.wso2.identity.organization.mgt.imp.pkg.version.range}",
                             org.wso2.carbon.identity.event.handler; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.event.event; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.event; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.event.services; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.role.v2.mgt.core.*;
                             version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon; version="${carbon.kernel.package.import.version.range}",
+                            *;resolution:=optional,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/pom.xml:226]: remove before merging, by adding all required imports

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 4]
```diff
@@ -42,17 +50,152 @@ public interface OrgApplicationManager {
...
+     * - SELECTED_ORG_ONLY: Shares with specified organizations only
+     * - SELECTED_ORG_WITH_ALL_EXISTING_AND_FUTURE_CHILDREN: Shares with specified organizations and their existing
+     * and future children
+     * <p>
+     * Features:
+     * - Share with multiple organizations in a single operation.
+     * - Configure roles that are shared with the organizations.
+     * - Automatically enforces parent-child inheritance rules.
+     *
+     * @param mainOrganizationId The ID of the organization that owns the application
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:71]:
```suggestion
     * @param mainOrganizationId The ID of the organization that owns the application.
```

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:71]: Fix other places as well

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 5]
```diff
@@ -42,17 +50,152 @@ public interface OrgApplicationManager {
...
+     * @param mainApplicationId The ID of the application to be shared
+     * @param selectiveShareApplicationList List of organizations to share with, including their policies and roles.
+     *                                      Important: You must include the complete organization hierarchy in this
+     *                                      list. For example, to share with organization A.1.1, you must include
+     *                                      both A.1 and A.1.1 in the list (order doesn't matter).
+     * @throws OrganizationManagementException If any error occurs during the sharing process
+     */
+    default void selectiveShareOrganizationApplication(String mainOrganizationId, String mainApplicationId,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:79]:
```suggestion
    default void selectiveApplicationShare(String mainOrganizationId, String mainApplicationId,
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 6]
```diff
@@ -42,17 +50,152 @@ public interface OrgApplicationManager {
...
+     * @param mainOrganizationId ID of the organization that owns the application.
+     * @param mainApplicationId ID of the application to be shared.
+     * @param generalApplicationShare Contains the sharing policy and role configuration that will be
+     *                               applied to all child organizations. Available policies:
+     *                               - `ALL_EXISTING_ORGS_ONLY` - Share with existing child organizations only.
+     *                               - `ALL_EXISTING_AND_FUTURE_ORGS` - Share with all existing and future child
+     *                                organizations.
+     * @throws OrganizationManagementException If the sharing operation fails.
+     */
+    default void generalOrganizationApplicationShare(String mainOrganizationId, String mainApplicationId,
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:103]:
```suggestion
    default void generalApplicationShare(String mainOrganizationId, String mainApplicationId,
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 7]
```diff
@@ -42,17 +50,152 @@ public interface OrgApplicationManager {
...
+                                                     GeneralApplicationShareOperation generalApplicationShare)
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:104]:
```suggestion
                                                     GeneralApplicationShareOperation generalApplicationShareOperation)
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationScimFilterParser.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+/**
+ * Utility class to parse SCIM-like filter strings for organization management.
+ * This class is specifically designed to handle filters in the format:
+ * - organizations[orgId eq "<orgIdValue>"]
+ * - organizations[orgId eq "<orgIdValue>"].roles
+ * <p>
+ * The class enforces a strict format and provides a method to extract the organization ID
+ * and an optional path attribute (e.g., ".roles").
+ */
+public class OrgApplicationScimFilterParser {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationScimFilterParser.java:36]: change the class name excluding SCIM, it's fine to have in the comment

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationManagerUtil.java [hunk 9]
```diff
@@ -218,4 +268,23 @@ private static ServiceProvider getDefaultServiceProvider() throws OrganizationMa
             throw new OrganizationManagementServerException("Error while retrieving default service provider", null, e);
         }
     }
+
+    public static OrganizationManagementClientException handleClientException(OrgApplicationMgtConstants.ErrorMessages
+                                                                                      error) {
+        return new OrganizationManagementClientException(error.getMessage(), error.getDescription(), error.getCode());
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationManagerUtil.java:274]: missing new line after method signature

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 10]
```diff
@@ -21,9 +21,15 @@
 import org.wso2.carbon.identity.application.common.model.ApplicationBasicInfo;
 import org.wso2.carbon.identity.application.common.model.ServiceProvider;
 import org.wso2.carbon.identity.organization.management.application.model.SharedApplication;
+import org.wso2.carbon.identity.organization.management.application.model.SharedApplicationOrganizationNodePage;
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:24]: Change the year range in copy right header.
Check other classes as well

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java [hunk 11]
```diff
@@ -42,17 +50,156 @@ public interface OrgApplicationManager {
...
+     *                               - `ALL_EXISTING_AND_FUTURE_ORGS` - Share with all existing and future child
+     *                                organizations.
+     * @throws OrganizationManagementException If the sharing operation fails.
+     */
+    default void shareApplicationWithAllOrganizations(String mainOrganizationId, String mainApplicationId,
+                                                     GeneralApplicationShareOperation generalApplicationShare)
+            throws OrganizationManagementException {
+
+        throw new NotImplementedException(
+                "generalOrganizationApplicationShare method is not implemented in " + this.getClass().getName());
```
Comment (Yasasr1) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManager.java:107]:
```suggestion
                "shareApplicationWithAllOrganizations method is not implemented in " + this.getClass().getName());
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 12]
```diff
@@ -660,57 +1462,207 @@ private boolean isOrganizationLoginIDP(IdentityProvider idp) {
...
+            String sharedApplicationId;
+            // Create the OAuth app and the service provider only if the app hasn't been created.
+            if (!mayBeSharedAppId.isPresent()) {
+                // Check if the application is shared to the parentOrg.
+                String parentOrgId = getOrganizationManager().getOrganization(sharingOrgId, false, false).getParent()
+                        .getId();
+                if (!parentOrgId.equals(ownerOrgId)) {
+                    Optional<String> parentOrgSharedAppId = resolveSharedApp(
+                            mainApplication.getApplicationResourceId(), ownerOrgId, parentOrgId);
+                    if (!parentOrgSharedAppId.isPresent()) {
```
Comment (Yasasr1) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:1520]: Is it possible for an application to be shared with the immediate parent organization, but not with higher-level parent organizations in the hierarchy?

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/OrganizationCreationHandler.java [hunk 13]
```diff
@@ -105,18 +122,150 @@ public void handleEvent(Event event) throws IdentityEventException {
...
+            String sharedAppId;
+            if (mainOrganizationId.equals(parentOrgId)) {
+                // That mean original application is available in the parent org. Not a fragment application.
+                sharedAppId = mainApplicationId;
+            } else {
+                Optional<String> sharedAppIdOptional = resolveSharedApp(mainApplicationId, mainOrganizationId,
+                        parentOrgId);
+                if (!sharedAppIdOptional.isPresent()) {
+                    // No shared application found for the main application.
+                    // TODO: Should we throw an error here and fail org creation?
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/OrganizationCreationHandler.java:163]: fix this todo

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/listener/OrganizationCreationHandler.java:163]: there are few more todos in the code

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/constant/OrgApplicationMgtConstants.java [hunk 14]
```diff
@@ -112,4 +157,74 @@ public String getValue() {
...
+                "Role sharing mode is invalid. Ensure that the role sharing policy is " +
+                        "either ALL, NONE or SELECTED."),
+        ERROR_CODE_INVALID_ROLE_SHARING_OPERATION("60014",
+                "Invalid role sharing operation.",
+                "Role sharing operation is invalid. Ensure that the role sharing operation is provided with valid " +
+                        "role sharing mode and role with audience list.");
+
+
+
+
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/constant/OrgApplicationMgtConstants.java:203]: remove unwanted newlines

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 15]
```diff
@@ -205,52 +254,163 @@ public void shareOrganizationApplication(String ownerOrgId, String originalAppId
...
+        Set<String> subOrganizationIdsSet = new HashSet<>(getOrganizationIdsInBfsOrder(childOrganizationGraph));
+
+        if (childOrganizationGraph.isEmpty()) {
+            return;
+        }
+
+        // Filter only valid child org configs that exist in the owner's child orgs using the map (O(1) lookups).
+        List<SelectiveShareApplicationOperation> filteredChildOrgConfigs = selectiveShareApplicationList.stream()
+                .filter(config -> subOrganizationIdsSet.contains(config.getOrganizationId()))
+                .collect(Collectors.toList());
```
Comment (HasiniSama) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:314]: It's better if we can add a debug log for invalid child org IDs sent.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java [hunk 16]
```diff
@@ -0,0 +1,533 @@
...
+            return new ArrayList<>(); // No valid organizations, return empty list.
+        }
+
+        // 2. Build parent-child relationship map for ancestry checking.
+        Map<String, String> parentMap = buildParentMap(allValidNodes);
+
+        // 3. Collect valid organization IDs and validate complete hierarchy paths.
+        Set<String> availableOrgIds = new HashSet<>(selectiveShareApplicationList.size());
+        Map<String, SelectiveShareApplicationOperation> orgIdToConfigMap = new HashMap<>(
+                selectiveShareApplicationList.size());
```
Comment (HasiniSama) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java:180]: Instead of using a set for **availableOrgIds** we can use, **orgIdToConfigMap.keySet()** which returns the list of available organization IDs.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+            processedOrgIds.add(descendantId);
+
+            // Only create inherited config if not already in effectiveConfigs
+            if (!effectiveConfigs.containsKey(descendantId)) {
+                SelectiveShareApplicationOperation inheritedConfig = new SelectiveShareApplicationOperation(
+                    descendantId,
+                    parentConfig.getPolicy(),
+                    parentConfig.getRoleSharing()
+                );
+                effectiveConfigs.put(descendantId, inheritedConfig);
```
Comment (HasiniSama) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java:268]: How can there be **descendantId** in the **effectiveConfigs** if we are iterating the Ids in the **bfsOrder**?

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java (hunk 16 above)
Comment (HasiniSama) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/util/OrgApplicationShareProcessor.java:180]: Map.keySet() returns a live view of the map’s keys. It's backed directly by the map — no copying or new allocation.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 18]
```diff
@@ -736,75 +1689,75 @@ private void handleShareApplicationException(String operationId, OrganizationMan
...
+//                    sharedApplicationId, sharedOrgId, shareWithAllChildren);
+//        } catch (IdentityApplicationManagementException e) {
+//            removeOAuthApplication(createdOAuthApp);
+//            processUnitOperationStatus(operationId, mainApplication.getApplicationResourceId(), sharedOrgId,
+//                    OperationStatus.FAILED, e.getMessage());
+//            throw handleServerException(ERROR_CODE_ERROR_SHARING_APPLICATION, e,
+//                    mainApplication.getApplicationResourceId(), sharedOrgId);
+//        }
+//        return sharedApplicationId;
+//    }
```
Comment (HasiniSama) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:1760]: Was there a particular reason this part is commented out?
